pkg/controller/bastion/actuator_delete.go

```diff
@@ -0,0 +1,199 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	azureclient "github.com/gardener/gardener-extension-provider-azure/pkg/azure/client"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	ctrlerror "github.com/gardener/gardener/extensions/pkg/controller/error"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (a *actuator) Delete(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "delete")
+
+	var factory = azureclient.NewAzureClientFactory(a.client)
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	err = removeBastionInstance(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove bastion instance: %w", err)
+	}
+
+	deleted, err := isInstanceDeleted(ctx, factory, opt)
+	if err != nil {
+		return fmt.Errorf("failed to check for bastion instance: %w", err)
+	}
+
+	if !deleted {
+		return &ctrlerror.RequeueAfterError{
+			RequeueAfter: 10 * time.Second,
+			Cause:        errors.New("bastion instance is still deleting"),
+		}
+	}
+
+	err = removeNic(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove nic: %w", err)
+	}
+
+	err = removePublicIP(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove public ip: %w", err)
+	}
+
+	err = removeDisk(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove disk: %w", err)
+	}
+
+	err = removeNSGRule(ctx, factory, logger, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove nsg rules: %w", err)
+	}
+
+	return nil
+}
+
+func removeNSGRule(ctx context.Context, factory azureclient.Factory, logger logr.Logger, opt *Options) error {
+	securityGroupResp, err := getNetworkSecurityGroup(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+
+	ruleList := []string{NSGIngressAllowSSHResourceName(opt.BastionInstanceName), NSGEgressDenyAllResourceName(opt.BastionInstanceName), NSGEgressAllowOnlyResourceName(opt.BastionInstanceName)}
+	for _, name := range ruleList {
+		*securityGroupResp.SecurityRules = removeSecurityRuleBody(name, *securityGroupResp.SecurityRules)
```

Comment:
fixed https://github.com/gardener/gardener-extension-provider-azure/pull/397/commits/e3662ac08a51f7175c4ed11319868e52c6b7a4d0

Here I assume that `securityGroupResp != nil` because when receiving it we also check for `err`  and if there is no error then we assume all fine. Otherwise, we have to check for nils all over the place...
I will revalidate other places as well.